commcare_connect/reports/helpers.py

```diff
@@ -17,87 +24,147 @@ def get_table_data_for_year_month(
+        payment_query.filter(**filter_kwargs, confirmed=True)
+        .values("opportunity_access__opportunity__delivery_type__name", "opportunity_access__user_id")
+        .annotate(approved_sum=Sum("amount_usd", default=0))
+    )
+    delivery_type_grouped_users = defaultdict(set)
+    for item in avg_top_flw_amount_paid:
+        delivery_type_grouped_users[item["opportunity_access__opportunity__delivery_type__name"]].add(
+            (item["opportunity_access__user_id"], item["approved_sum"])
+        )
+    avg_top_flw_amount_paid_data = {}
+    for d_name, users in delivery_type_grouped_users.items():
+        sum_total_users = defaultdict(int)
+        for user, amount in users:
+            sum_total_users[user] += amount
+
+        flw_amount_paid_data[d_name] = sum(sum_total_users.values())
+        # take atleast 1 top user in cases where this variable is 0
+        top_five_percent_len = len(sum_total_users) // 20 or 1
+        avg_top_flw_amount_paid_data[d_name] = sum(
+            sorted(sum_total_users.values(), reverse=True)[:top_five_percent_len]
+        )
+
+    if group_by_delivery_type:
+        for delivery_type_name in user_count_data.keys():
+            nm_amount_earned = nm_amount_earned_data.get(delivery_type_name, 0)
+            nm_amount_paid = nm_amount_paid_data.get(delivery_type_name, 0)
+            data.append(
                 {
-                    "approved_payments": sum(approved_payment_dict.values()),
-                    "total_payments": sum(total_payment_dict.values()),
-                    "beneficiaries": len(beneficiary_set[delivery_type_name]),
+                    "delivery_type": delivery_type_name,
+                    "month": (month, year),
+                    "users": user_count_data[delivery_type_name],
+                    "services": service_count_data[delivery_type_name],
+                    "avg_time_to_payment": avg_time_to_payment_data.get(delivery_type_name),
+                    "max_time_to_payment": max_time_to_payment_data.get(delivery_type_name),
+                    "flw_amount_earned": flw_amount_earned_data.get(delivery_type_name, 0),
+                    "flw_amount_paid": flw_amount_paid_data.get(delivery_type_name, 0),
+                    "nm_amount_earned": nm_amount_earned,
+                    "nm_amount_paid": nm_amount_paid,
+                    "avg_top_paid_flws": avg_top_flw_amount_paid_data.get(delivery_type_name, 0),
                 }
             )
-        data.append(delivery_type_data)
+
+    else:
+        nm_amount_earned = sum(nm_amount_earned_data.values())
+        nm_amount_paid = sum(nm_amount_paid_data.values())
+        data.append(
+            {
+                "delivery_type": "All",
+                "month": (month, year),
+                "users": sum(user_count_data.values()),
+                "services": sum(service_count_data.values()),
+                "avg_time_to_payment": mean(avg_time_to_payment_data.values() or [0]),
```

Comment:
Wouldn't all of these be faster if these all aggs are done within DB when `group_by_delivery_type` is not True?